File: array/941.java
class Solution {
    public boolean validMountainArray(int[] A) {
        if (A.length < 3) {
            return false;
        }
        int pivot = 0;
        boolean inc = false;
        for (int i = 0; i < A.length-1; i++) {
            if (A[i] < A[i+1]) {
                inc = true;
            } else {
                pivot = i;
                break;
            }
        }
        for (int i = pivot; i < A.length-1; i++) {
            if (A[i] <= A[i+1]) {
                return false;
            }
        }
        if (!inc) {
            return false;
        }
        return true;
    }
}

File: array/941.cc
class Solution {
public:
    bool validMountainArray(vector<int>& arr) {
        if (arr.size() < 3) return false;
        int i = 0, j = arr.size()-1;
        // first half
        while (i < (arr.size()-1) && arr[i] < arr[i+1])
            ++i;
        // second half
        while (j > 0 && arr[j] < arr[j-1])
            --j;
        if (i == 0 || j == (arr.size()-1)) return false;
        return i == j;
    }
};
